README.md
# Go API client for OpalStack


## The Opalstack JSON REST API

### Authorization
This API uses an **Authorization** header of the form: `\"Authorization: Token 1111111111111111111111111111111111111111\"`, where **1111111111111111111111111111111111111111** represents an API token created at https://my.opalstack.com/tokens/.

The typical format of an API request looks like the following:
```
GET request:
    curl -s -H \"Authorization: Token 1111111111111111111111111111111111111111\" \"https://my.opalstack.com/api/v1/site/list/\" | jq .

POST request:
    curl -s -H \"Content-Type: application/json\" -H \"Authorization: Token 1111111111111111111111111111111111111111\" \\
            -X POST -d '[{\"id\": \"(site UUID)\", \"redirect\": true, ...}]' \"https://my.opalstack.com/api/v1/site/update/\" | jq .
```
(Further examples will omit **headers** and **jq** for the sake of clarity)

You can also authorize requests on our API Documentation page (https://my.opalstack.com/api/v1/doc/) in order to facilitate development.
To do so, click the \"**Authorize**\" button on the right side of the page and enter \"**Token 1111111111111111111111111111111111111111**\" in the **Value** field within.
Afterword, you will be able to perform requests directly from the documentation page. Be sure to logout when finished.

### Embedding
The Opalstack API supports _embedding_. This allows you to nest child API objects in a single GET request. For example, consider the following GET request performed with **curl**:
```
Request:
    curl \"https://my.opalstack.com/api/v1/osuser/list/\"

Response:
    [
      {
        \"id\": \"01010101-0202-0303-0404-050505050505\",
        \"state\": \"READY\",
        \"ready\": true,
        \"name\": \"the_osuser_name\",
        \"server\": \"11111111-1212-1313-1414-151515151515\"
      }
    ]
```

Suppose then that we would like additional information about the **server**. We _could_ proceed to query the **server** UUID (**11111111-1212-1313-1414-151515151515**) at the `/server/read/{uuid}` endpoint, like this:
```
Request:
    curl \"https://my.opalstack.com/api/v1/server/read/11111111-1212-1313-1414-151515151515\"

Response:
    {
      \"id\": \"11111111-1212-1313-1414-151515151515\",
      \"hostname\": \"vpsNNN.opalstack.com\"
    }
```

However, we could have instead choosen to specify `?embed=server` as a query parameter to the original GET request. This will cause objects to be _embedded_ in the response directly:
```
Request:
    curl \"https://my.opalstack.com/api/v1/osuser/list/?embed=server\"

Response:
    [
      {
        \"id\": \"01010101-0202-0303-0404-050505050505\",
        \"state\": \"READY\",
        \"ready\": true,
        \"name\": \"the_osuser_name\",
        \"server\": {
          \"id\": \"11111111-1212-1313-1414-151515151515\",
          \"hostname\": \"vpsNNN.opalstack.com\"
        }
      }
    ]
```
Here, the **server** field has been be populated with the full object instead of just a UUID.

The **embed** query parameter accepts multiple (comma-separated) fields to embed. For example:
```
curl \"https://my.opalstack.com/api/v1/account/info/?embed=web_servers,imap_servers,smtp_servers\"
```


## Overview
This API client was generated by the [OpenAPI Generator](https://openapi-generator.tech) project.  By using the [OpenAPI-spec](https://www.openapis.org/) from a remote server, you can easily generate an API client.

- API version: 1.0.0
- Package version: 0.0.1
- Build package: org.openapitools.codegen.languages.GoClientCodegen

## Installation

Install the following dependencies:

```shell
go get github.com/stretchr/testify/assert
go get golang.org/x/net/context
```

Put the package under your project folder and add the following in import:

```golang
import OpalStack "github.com/vlazic/opalstack-go"
```

To use a proxy, set the environment variable `HTTP_PROXY`:

```golang
os.Setenv("HTTP_PROXY", "http://proxy_name:proxy_port")
```

## Configuration of Server URL

Default configuration comes with `Servers` field that contains server objects as defined in the OpenAPI specification.

### Select Server Configuration

For using other server than the one defined on index 0 set context value `sw.ContextServerIndex` of type `int`.

```golang
ctx := context.WithValue(context.Background(), OpalStack.ContextServerIndex, 1)
```

### Templated Server URL

Templated server URL is formatted using default variables from configuration or from context value `sw.ContextServerVariables` of type `map[string]string`.

```golang
ctx := context.WithValue(context.Background(), OpalStack.ContextServerVariables, map[string]string{
	"basePath": "v2",
})
```

Note, enum values are always validated and all unused variables are silently ignored.

### URLs Configuration per Operation

Each operation can use different server URL defined using `OperationServers` map in the `Configuration`.
An operation is uniquely identified by `"{classname}Service.{nickname}"` string.
Similar rules for overriding default operation server index and variables applies by using `sw.ContextOperationServerIndices` and `sw.ContextOperationServerVariables` context maps.

```golang
ctx := context.WithValue(context.Background(), OpalStack.ContextOperationServerIndices, map[string]int{
	"{classname}Service.{nickname}": 2,
})
ctx = context.WithValue(context.Background(), OpalStack.ContextOperationServerVariables, map[string]map[string]string{
	"{classname}Service.{nickname}": {
		"port": "8443",
	},
})
```

## Documentation for API Endpoints

All URIs are relative to *http://localhost*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*AccountAPI* | [**AccountList**](docs/AccountAPI.md#accountlist) | **Get** /api/v1/account/list/ | 
*AccountAPI* | [**AccountRead**](docs/AccountAPI.md#accountread) | **Get** /api/v1/account/read/{uuid} | 
*AccountAPI* | [**AccountReadCurrent**](docs/AccountAPI.md#accountreadcurrent) | **Get** /api/v1/account/info/ | 
*AccountAPI* | [**AccountReadCurrent2**](docs/AccountAPI.md#accountreadcurrent2) | **Get** /api/v1/account/read/current/ | 
*AccountAPI* | [**AccountUpdate**](docs/AccountAPI.md#accountupdate) | **Post** /api/v1/account/update/ | 
*AccountAPI* | [**AccountUpdateCurrent**](docs/AccountAPI.md#accountupdatecurrent) | **Post** /api/v1/account/update/current/ | 
*AddressAPI* | [**AddressCreate**](docs/AddressAPI.md#addresscreate) | **Post** /api/v1/address/create/ | 
*AddressAPI* | [**AddressDelete**](docs/AddressAPI.md#addressdelete) | **Post** /api/v1/address/delete/ | 
*AddressAPI* | [**AddressList**](docs/AddressAPI.md#addresslist) | **Get** /api/v1/address/list/ | 
*AddressAPI* | [**AddressRead**](docs/AddressAPI.md#addressread) | **Get** /api/v1/address/read/{uuid} | 
*AddressAPI* | [**AddressUpdate**](docs/AddressAPI.md#addressupdate) | **Post** /api/v1/address/update/ | 
*AppAPI* | [**AppCreate**](docs/AppAPI.md#appcreate) | **Post** /api/v1/app/create/ | 
*AppAPI* | [**AppDelete**](docs/AppAPI.md#appdelete) | **Post** /api/v1/app/delete/ | 
*AppAPI* | [**AppFailed**](docs/AppAPI.md#appfailed) | **Post** /api/v1/app/failed/ | 
*AppAPI* | [**AppInstalled**](docs/AppAPI.md#appinstalled) | **Post** /api/v1/app/installed/ | 
*AppAPI* | [**AppList**](docs/AppAPI.md#applist) | **Get** /api/v1/app/list/ | 
*AppAPI* | [**AppRead**](docs/AppAPI.md#appread) | **Get** /api/v1/app/read/{uuid} | 
*AppAPI* | [**AppUpdate**](docs/AppAPI.md#appupdate) | **Post** /api/v1/app/update/ | 
*CertAPI* | [**CertCreate**](docs/CertAPI.md#certcreate) | **Post** /api/v1/cert/create/ | 
*CertAPI* | [**CertDelete**](docs/CertAPI.md#certdelete) | **Post** /api/v1/cert/delete/ | 
*CertAPI* | [**CertList**](docs/CertAPI.md#certlist) | **Get** /api/v1/cert/list/ | 
*CertAPI* | [**CertRead**](docs/CertAPI.md#certread) | **Get** /api/v1/cert/read/{uuid} | 
*CertAPI* | [**CertShared**](docs/CertAPI.md#certshared) | **Get** /api/v1/cert/shared/ | 

CHANGELOG.md
# 1.0.0 (2023-08-29)


### Bug Fixes

*  prefix enum with class name to avoid same enum keys ([d05b5fa](https://github.com/vlazic/opalstack-go/commit/d05b5fa076062a873d2067c2ab5723b5e348bc45))
* fix location of openapiclient package ([8a1c1cd](https://github.com/vlazic/opalstack-go/commit/8a1c1cd86133c1591701fe7ba03223c32769224b))


### Features

* add goreleaser github action; update dependencies ([ab14d0b](https://github.com/vlazic/opalstack-go/commit/ab14d0ba7c84cab1ade0f34306e0ed9daad6c8a8))
* update generate script and references to github repo ([f2b699c](https://github.com/vlazic/opalstack-go/commit/f2b699c957fde1d46ec2d879c1a1ba2413c10e74))


api_osvar.go
/*
Opalstack API

 ## The Opalstack JSON REST API  ### Authorization This API uses an **Authorization** header of the form: `\"Authorization: Token 1111111111111111111111111111111111111111\"`, where **1111111111111111111111111111111111111111** represents an API token created at https://my.opalstack.com/tokens/.  The typical format of an API request looks like the following: ``` GET request:     curl -s -H \"Authorization: Token 1111111111111111111111111111111111111111\" \"https://my.opalstack.com/api/v1/site/list/\" | jq .  POST request:     curl -s -H \"Content-Type: application/json\" -H \"Authorization: Token 1111111111111111111111111111111111111111\" \\             -X POST -d '[{\"id\": \"(site UUID)\", \"redirect\": true, ...}]' \"https://my.opalstack.com/api/v1/site/update/\" | jq . ``` (Further examples will omit **headers** and **jq** for the sake of clarity)  You can also authorize requests on our API Documentation page (https://my.opalstack.com/api/v1/doc/) in order to facilitate development. To do so, click the \"**Authorize**\" button on the right side of the page and enter \"**Token 1111111111111111111111111111111111111111**\" in the **Value** field within. Afterword, you will be able to perform requests directly from the documentation page. Be sure to logout when finished.  ### Embedding The Opalstack API supports _embedding_. This allows you to nest child API objects in a single GET request. For example, consider the following GET request performed with **curl**: ``` Request:     curl \"https://my.opalstack.com/api/v1/osuser/list/\"  Response:     [       {         \"id\": \"01010101-0202-0303-0404-050505050505\",         \"state\": \"READY\",         \"ready\": true,         \"name\": \"the_osuser_name\",         \"server\": \"11111111-1212-1313-1414-151515151515\"       }     ] ```  Suppose then that we would like additional information about the **server**. We _could_ proceed to query the **server** UUID (**11111111-1212-1313-1414-151515151515**) at the `/server/read/{uuid}` endpoint, like this: ``` Request:     curl \"https://my.opalstack.com/api/v1/server/read/11111111-1212-1313-1414-151515151515\"  Response:     {       \"id\": \"11111111-1212-1313-1414-151515151515\",       \"hostname\": \"vpsNNN.opalstack.com\"     } ```  However, we could have instead choosen to specify `?embed=server` as a query parameter to the original GET request. This will cause objects to be _embedded_ in the response directly: ``` Request:     curl \"https://my.opalstack.com/api/v1/osuser/list/?embed=server\"  Response:     [       {         \"id\": \"01010101-0202-0303-0404-050505050505\",         \"state\": \"READY\",         \"ready\": true,         \"name\": \"the_osuser_name\",         \"server\": {           \"id\": \"11111111-1212-1313-1414-151515151515\",           \"hostname\": \"vpsNNN.opalstack.com\"         }       }     ] ``` Here, the **server** field has been be populated with the full object instead of just a UUID.  The **embed** query parameter accepts multiple (comma-separated) fields to embed. For example: ``` curl \"https://my.opalstack.com/api/v1/account/info/?embed=web_servers,imap_servers,smtp_servers\" ``` 

API version: 1.0.0
*/

// Code generated by OpenAPI Generator (https://openapi-generator.tech); DO NOT EDIT.

package OpalStack

import (
	"bytes"
	"context"
	"io"
	"net/http"
	"net/url"
	"strings"
)


// OsvarAPIService OsvarAPI service
type OsvarAPIService service

type ApiOsvarCreateRequest struct {
	ctx context.Context
	ApiService *OsvarAPIService
	oSVarCreate *[]OSVarCreate
}

func (r ApiOsvarCreateRequest) OSVarCreate(oSVarCreate []OSVarCreate) ApiOsvarCreateRequest {
	r.oSVarCreate = &oSVarCreate
	return r
}

func (r ApiOsvarCreateRequest) Execute() ([]OSVarResponse, *http.Response, error) {
	return r.ApiService.OsvarCreateExecute(r)
}

/*
OsvarCreate Method for OsvarCreate

 @param ctx context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
 @return ApiOsvarCreateRequest
*/
func (a *OsvarAPIService) OsvarCreate(ctx context.Context) ApiOsvarCreateRequest {
	return ApiOsvarCreateRequest{
		ApiService: a,
		ctx: ctx,
	}
}

// Execute executes the request
//  @return []OSVarResponse
func (a *OsvarAPIService) OsvarCreateExecute(r ApiOsvarCreateRequest) ([]OSVarResponse, *http.Response, error) {
	var (
		localVarHTTPMethod   = http.MethodPost
		localVarPostBody     interface{}
		formFiles            []formFile
		localVarReturnValue  []OSVarResponse
	)

	localBasePath, err := a.client.cfg.ServerURLWithContext(r.ctx, "OsvarAPIService.OsvarCreate")
	if err != nil {
		return localVarReturnValue, nil, &GenericOpenAPIError{error: err.Error()}
	}

	localVarPath := localBasePath + "/api/v1/osvar/create/"

	localVarHeaderParams := make(map[string]string)
	localVarQueryParams := url.Values{}
	localVarFormParams := url.Values{}
	if r.oSVarCreate == nil {
		return localVarReturnValue, nil, reportError("oSVarCreate is required and must be specified")
	}

	// to determine the Content-Type header
	localVarHTTPContentTypes := []string{"application/json", "application/x-www-form-urlencoded"}

	// set Content-Type header
	localVarHTTPContentType := selectHeaderContentType(localVarHTTPContentTypes)
	if localVarHTTPContentType != "" {
		localVarHeaderParams["Content-Type"] = localVarHTTPContentType
	}

	// to determine the Accept header
	localVarHTTPHeaderAccepts := []string{"application/json"}

	// set Accept header
	localVarHTTPHeaderAccept := selectHeaderAccept(localVarHTTPHeaderAccepts)
	if localVarHTTPHeaderAccept != "" {
		localVarHeaderParams["Accept"] = localVarHTTPHeaderAccept
	}
	// body params
	localVarPostBody = r.oSVarCreate
	if r.ctx != nil {
		// API Key Authentication
		if auth, ok := r.ctx.Value(ContextAPIKeys).(map[string]APIKey); ok {
			if apiKey, ok := auth["tokenAuth"]; ok {
				var key string
				if apiKey.Prefix != "" {
					key = apiKey.Prefix + " " + apiKey.Key
				} else {
					key = apiKey.Key
				}
				localVarHeaderParams["Authorization"] = key
			}
		}
	}
	req, err := a.client.prepareRequest(r.ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, formFiles)
	if err != nil {
		return localVarReturnValue, nil, err
	}

	localVarHTTPResponse, err := a.client.callAPI(req)
	if err != nil || localVarHTTPResponse == nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	localVarBody, err := io.ReadAll(localVarHTTPResponse.Body)
	localVarHTTPResponse.Body.Close()
	localVarHTTPResponse.Body = io.NopCloser(bytes.NewBuffer(localVarBody))
	if err != nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	if localVarHTTPResponse.StatusCode >= 300 {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: localVarHTTPResponse.Status,
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	err = a.client.decode(&localVarReturnValue, localVarBody, localVarHTTPResponse.Header.Get("Content-Type"))
	if err != nil {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: err.Error(),
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	return localVarReturnValue, localVarHTTPResponse, nil
}

type ApiOsvarDeleteRequest struct {
	ctx context.Context
	ApiService *OsvarAPIService
	oSVarRead *[]OSVarRead
}

func (r ApiOsvarDeleteRequest) OSVarRead(oSVarRead []OSVarRead) ApiOsvarDeleteRequest {
	r.oSVarRead = &oSVarRead
	return r
}

func (r ApiOsvarDeleteRequest) Execute() (*http.Response, error) {
	return r.ApiService.OsvarDeleteExecute(r)
}

/*
OsvarDelete Method for OsvarDelete

 @param ctx context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
 @return ApiOsvarDeleteRequest
*/
func (a *OsvarAPIService) OsvarDelete(ctx context.Context) ApiOsvarDeleteRequest {
	return ApiOsvarDeleteRequest{
		ApiService: a,
		ctx: ctx,
	}
}

// Execute executes the request
func (a *OsvarAPIService) OsvarDeleteExecute(r ApiOsvarDeleteRequest) (*http.Response, error) {
	var (
		localVarHTTPMethod   = http.MethodPost
		localVarPostBody     interface{}
		formFiles            []formFile
	)

	localBasePath, err := a.client.cfg.ServerURLWithContext(r.ctx, "OsvarAPIService.OsvarDelete")
	if err != nil {
		return nil, &GenericOpenAPIError{error: err.Error()}
	}

	localVarPath := localBasePath + "/api/v1/osvar/delete/"

	localVarHeaderParams := make(map[string]string)
	localVarQueryParams := url.Values{}
	localVarFormParams := url.Values{}
	if r.oSVarRead == nil {
		return nil, reportError("oSVarRead is required and must be specified")
	}

	// to determine the Content-Type header
	localVarHTTPContentTypes := []string{"application/json", "application/x-www-form-urlencoded"}

	// set Content-Type header
	localVarHTTPContentType := selectHeaderContentType(localVarHTTPContentTypes)
	if localVarHTTPContentType != "" {
		localVarHeaderParams["Content-Type"] = localVarHTTPContentType
	}

	// to determine the Accept header
	localVarHTTPHeaderAccepts := []string{}

	// set Accept header
	localVarHTTPHeaderAccept := selectHeaderAccept(localVarHTTPHeaderAccepts)
	if localVarHTTPHeaderAccept != "" {
		localVarHeaderParams["Accept"] = localVarHTTPHeaderAccept
	}
	// body params
	localVarPostBody = r.oSVarRead
	if r.ctx != nil {
		// API Key Authentication
		if auth, ok := r.ctx.Value(ContextAPIKeys).(map[string]APIKey); ok {
			if apiKey, ok := auth["tokenAuth"]; ok {
				var key string
				if apiKey.Prefix != "" {
					key = apiKey.Prefix + " " + apiKey.Key
				} else {
					key = apiKey.Key
				}
				localVarHeaderParams["Authorization"] = key
			}
		}
	}
	req, err := a.client.prepareRequest(r.ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, formFiles)
	if err != nil {
		return nil, err
	}

	localVarHTTPResponse, err := a.client.callAPI(req)
	if err != nil || localVarHTTPResponse == nil {
		return localVarHTTPResponse, err
	}

	localVarBody, err := io.ReadAll(localVarHTTPResponse.Body)
	localVarHTTPResponse.Body.Close()
	localVarHTTPResponse.Body = io.NopCloser(bytes.NewBuffer(localVarBody))
	if err != nil {
		return localVarHTTPResponse, err
	}

	if localVarHTTPResponse.StatusCode >= 300 {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: localVarHTTPResponse.Status,
		}
		return localVarHTTPResponse, newErr
	}

	return localVarHTTPResponse, nil
}

type ApiOsvarListRequest struct {
	ctx context.Context
	ApiService *OsvarAPIService
}

func (r ApiOsvarListRequest) Execute() ([]OSVarResponse, *http.Response, error) {
	return r.ApiService.OsvarListExecute(r)
}

/*
OsvarList Method for OsvarList

 @param ctx context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
 @return ApiOsvarListRequest
*/
func (a *OsvarAPIService) OsvarList(ctx context.Context) ApiOsvarListRequest {
	return ApiOsvarListRequest{
		ApiService: a,
		ctx: ctx,
	}
}

// Execute executes the request
//  @return []OSVarResponse
func (a *OsvarAPIService) OsvarListExecute(r ApiOsvarListRequest) ([]OSVarResponse, *http.Response, error) {
	var (
		localVarHTTPMethod   = http.MethodGet
		localVarPostBody     interface{}
		formFiles            []formFile
		localVarReturnValue  []OSVarResponse
	)

	localBasePath, err := a.client.cfg.ServerURLWithContext(r.ctx, "OsvarAPIService.OsvarList")
	if err != nil {
		return localVarReturnValue, nil, &GenericOpenAPIError{error: err.Error()}
	}

	localVarPath := localBasePath + "/api/v1/osvar/list/"

	localVarHeaderParams := make(map[string]string)
	localVarQueryParams := url.Values{}
	localVarFormParams := url.Values{}

	// to determine the Content-Type header
	localVarHTTPContentTypes := []string{}

	// set Content-Type header
	localVarHTTPContentType := selectHeaderContentType(localVarHTTPContentTypes)
	if localVarHTTPContentType != "" {
		localVarHeaderParams["Content-Type"] = localVarHTTPContentType
	}

	// to determine the Accept header
	localVarHTTPHeaderAccepts := []string{"application/json"}

	// set Accept header
	localVarHTTPHeaderAccept := selectHeaderAccept(localVarHTTPHeaderAccepts)
	if localVarHTTPHeaderAccept != "" {
		localVarHeaderParams["Accept"] = localVarHTTPHeaderAccept
	}
	if r.ctx != nil {
		// API Key Authentication
		if auth, ok := r.ctx.Value(ContextAPIKeys).(map[string]APIKey); ok {
			if apiKey, ok := auth["tokenAuth"]; ok {
				var key string
				if apiKey.Prefix != "" {
					key = apiKey.Prefix + " " + apiKey.Key
				} else {
					key = apiKey.Key
				}
				localVarHeaderParams["Authorization"] = key
			}
		}
	}
	req, err := a.client.prepareRequest(r.ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, formFiles)
	if err != nil {
		return localVarReturnValue, nil, err
	}

	localVarHTTPResponse, err := a.client.callAPI(req)
	if err != nil || localVarHTTPResponse == nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	localVarBody, err := io.ReadAll(localVarHTTPResponse.Body)
	localVarHTTPResponse.Body.Close()
	localVarHTTPResponse.Body = io.NopCloser(bytes.NewBuffer(localVarBody))
	if err != nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	if localVarHTTPResponse.StatusCode >= 300 {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: localVarHTTPResponse.Status,
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	err = a.client.decode(&localVarReturnValue, localVarBody, localVarHTTPResponse.Header.Get("Content-Type"))
	if err != nil {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: err.Error(),
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	return localVarReturnValue, localVarHTTPResponse, nil
}

type ApiOsvarReadRequest struct {
	ctx context.Context
	ApiService *OsvarAPIService
	uuid string
}

func (r ApiOsvarReadRequest) Execute() (*OSVarResponse, *http.Response, error) {
	return r.ApiService.OsvarReadExecute(r)
}

/*
OsvarRead Method for OsvarRead

 @param ctx context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
 @param uuid
 @return ApiOsvarReadRequest
*/
func (a *OsvarAPIService) OsvarRead(ctx context.Context, uuid string) ApiOsvarReadRequest {
	return ApiOsvarReadRequest{
		ApiService: a,
		ctx: ctx,
		uuid: uuid,
	}
}

// Execute executes the request
//  @return OSVarResponse
func (a *OsvarAPIService) OsvarReadExecute(r ApiOsvarReadRequest) (*OSVarResponse, *http.Response, error) {
	var (
		localVarHTTPMethod   = http.MethodGet
		localVarPostBody     interface{}
		formFiles            []formFile
		localVarReturnValue  *OSVarResponse
	)

	localBasePath, err := a.client.cfg.ServerURLWithContext(r.ctx, "OsvarAPIService.OsvarRead")
	if err != nil {
		return localVarReturnValue, nil, &GenericOpenAPIError{error: err.Error()}
	}

	localVarPath := localBasePath + "/api/v1/osvar/read/{uuid}"
	localVarPath = strings.Replace(localVarPath, "{"+"uuid"+"}", url.PathEscape(parameterValueToString(r.uuid, "uuid")), -1)

	localVarHeaderParams := make(map[string]string)
	localVarQueryParams := url.Values{}
	localVarFormParams := url.Values{}

	// to determine the Content-Type header
	localVarHTTPContentTypes := []string{}

	// set Content-Type header
	localVarHTTPContentType := selectHeaderContentType(localVarHTTPContentTypes)
	if localVarHTTPContentType != "" {
		localVarHeaderParams["Content-Type"] = localVarHTTPContentType
	}

	// to determine the Accept header
	localVarHTTPHeaderAccepts := []string{"application/json"}

	// set Accept header
	localVarHTTPHeaderAccept := selectHeaderAccept(localVarHTTPHeaderAccepts)
	if localVarHTTPHeaderAccept != "" {
		localVarHeaderParams["Accept"] = localVarHTTPHeaderAccept
	}
	if r.ctx != nil {
		// API Key Authentication
		if auth, ok := r.ctx.Value(ContextAPIKeys).(map[string]APIKey); ok {
			if apiKey, ok := auth["tokenAuth"]; ok {
				var key string
				if apiKey.Prefix != "" {
					key = apiKey.Prefix + " " + apiKey.Key
				} else {
					key = apiKey.Key
				}
				localVarHeaderParams["Authorization"] = key
			}
		}
	}
	req, err := a.client.prepareRequest(r.ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, formFiles)
	if err != nil {
		return localVarReturnValue, nil, err
	}

	localVarHTTPResponse, err := a.client.callAPI(req)
	if err != nil || localVarHTTPResponse == nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	localVarBody, err := io.ReadAll(localVarHTTPResponse.Body)
	localVarHTTPResponse.Body.Close()
	localVarHTTPResponse.Body = io.NopCloser(bytes.NewBuffer(localVarBody))
	if err != nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	if localVarHTTPResponse.StatusCode >= 300 {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: localVarHTTPResponse.Status,
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	err = a.client.decode(&localVarReturnValue, localVarBody, localVarHTTPResponse.Header.Get("Content-Type"))
	if err != nil {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: err.Error(),
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	return localVarReturnValue, localVarHTTPResponse, nil
}

type ApiOsvarUpdateRequest struct {
	ctx context.Context
	ApiService *OsvarAPIService
	oSVarUpdate *[]OSVarUpdate
}

func (r ApiOsvarUpdateRequest) OSVarUpdate(oSVarUpdate []OSVarUpdate) ApiOsvarUpdateRequest {
	r.oSVarUpdate = &oSVarUpdate
	return r
}

func (r ApiOsvarUpdateRequest) Execute() ([]OSVarResponse, *http.Response, error) {
	return r.ApiService.OsvarUpdateExecute(r)
}

/*
OsvarUpdate Method for OsvarUpdate

 @param ctx context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
 @return ApiOsvarUpdateRequest
*/
func (a *OsvarAPIService) OsvarUpdate(ctx context.Context) ApiOsvarUpdateRequest {
	return ApiOsvarUpdateRequest{
		ApiService: a,
		ctx: ctx,
	}
}

// Execute executes the request
//  @return []OSVarResponse
func (a *OsvarAPIService) OsvarUpdateExecute(r ApiOsvarUpdateRequest) ([]OSVarResponse, *http.Response, error) {
	var (
		localVarHTTPMethod   = http.MethodPost
		localVarPostBody     interface{}
		formFiles            []formFile
		localVarReturnValue  []OSVarResponse
	)

	localBasePath, err := a.client.cfg.ServerURLWithContext(r.ctx, "OsvarAPIService.OsvarUpdate")
	if err != nil {
		return localVarReturnValue, nil, &GenericOpenAPIError{error: err.Error()}
	}

	localVarPath := localBasePath + "/api/v1/osvar/update/"

	localVarHeaderParams := make(map[string]string)
	localVarQueryParams := url.Values{}
	localVarFormParams := url.Values{}
	if r.oSVarUpdate == nil {
		return localVarReturnValue, nil, reportError("oSVarUpdate is required and must be specified")
	}

	// to determine the Content-Type header
	localVarHTTPContentTypes := []string{"application/json", "application/x-www-form-urlencoded"}

	// set Content-Type header
	localVarHTTPContentType := selectHeaderContentType(localVarHTTPContentTypes)
	if localVarHTTPContentType != "" {
		localVarHeaderParams["Content-Type"] = localVarHTTPContentType
	}

	// to determine the Accept header
	localVarHTTPHeaderAccepts := []string{"application/json"}

	// set Accept header
	localVarHTTPHeaderAccept := selectHeaderAccept(localVarHTTPHeaderAccepts)
	if localVarHTTPHeaderAccept != "" {
		localVarHeaderParams["Accept"] = localVarHTTPHeaderAccept
	}
	// body params
	localVarPostBody = r.oSVarUpdate
	if r.ctx != nil {
		// API Key Authentication
		if auth, ok := r.ctx.Value(ContextAPIKeys).(map[string]APIKey); ok {
			if apiKey, ok := auth["tokenAuth"]; ok {
				var key string
				if apiKey.Prefix != "" {
					key = apiKey.Prefix + " " + apiKey.Key
				} else {
					key = apiKey.Key
				}
				localVarHeaderParams["Authorization"] = key
			}
		}
	}
	req, err := a.client.prepareRequest(r.ctx, localVarPath, localVarHTTPMethod, localVarPostBody, localVarHeaderParams, localVarQueryParams, localVarFormParams, formFiles)
	if err != nil {
		return localVarReturnValue, nil, err
	}

	localVarHTTPResponse, err := a.client.callAPI(req)
	if err != nil || localVarHTTPResponse == nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	localVarBody, err := io.ReadAll(localVarHTTPResponse.Body)
	localVarHTTPResponse.Body.Close()
	localVarHTTPResponse.Body = io.NopCloser(bytes.NewBuffer(localVarBody))
	if err != nil {
		return localVarReturnValue, localVarHTTPResponse, err
	}

	if localVarHTTPResponse.StatusCode >= 300 {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: localVarHTTPResponse.Status,
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	err = a.client.decode(&localVarReturnValue, localVarBody, localVarHTTPResponse.Header.Get("Content-Type"))
	if err != nil {
		newErr := &GenericOpenAPIError{
			body:  localVarBody,
			error: err.Error(),
		}
		return localVarReturnValue, localVarHTTPResponse, newErr
	}

	return localVarReturnValue, localVarHTTPResponse, nil
}
